More SW infrastructure

faisal-bhuiyan opened the pull request:
Following enhancements to the OpenTurbine SW infrastructure are proposed here:

- Create API documentation via Doxygen - Use of doxygen specific annotation embedded within C++ source code is pretty common to generate API documentation with minimal effort, e.g. in [AMR wind](https://exawind.github.io/amr-wind/developer/documentation.html#documenting-source-code).

- Adopt a version number for project in CMakeLists.txt - we might consider adopting a [version number](https://cmake.org/cmake/help/latest/variable/CMAKE_PROJECT_VERSION.html) for OpenTurbine using CMake. We could use the version number along with a log file (discussed below) to quickly check the build version for development as well as for releases.

- Create a logging utility - a logging utility for recording log messages to files (alongside the console) should be a very useful development tool and the first point of debugging whenever necessary. Various open-source libraries are available for this such as [spdlog](https://github.com/gabime/spdlog), however, we could just implement our own basic logging utility with minimal effort.

- Adopt a C++ Style guide - We are currently using the C++ style guide in the form of a `clang-format` as recommended in [amr-wind](https://github.com/Exawind/amr-wind/blob/main/.clang-format). However, having a descriptive style guide helps to know about the thought process behind the choices made. The [Google C++ style guide](https://google.github.io/styleguide/cppguide.html) might be a good candidate to this end, of course with a few modifications of our own, e.g., increasing the [line length](https://google.github.io/styleguide/cppguide.html#Line_Length) from 80 characters to a more reasonable value such as 120.

rafmudaf commented on CMakeLists.txt:
```diff
@@ -2,7 +2,7 @@
 
 cmake_minimum_required (VERSION 3.20 FATAL_ERROR)
-project(OpenTurbine CXX)
+project(OpenTurbine VERSION 23.0.1 LANGUAGES CXX)
```
We should probably start out with something like version 0.1.0. I've tagged the initial commit to v0.0.0 (as in, I added a git-tag). Also, keep in mind that the version is handled through https://github.com/Exawind/openturbine/blob/main/cmake/GetGitRevisionDescription.cmake so you'll have to keep these two in sync.

alanw0 replied:
@rafmudaf I really like the idea of using a git revision as the version, this is the most precise way to tie a run of the code back to the actual code that was built.
It would be nice if cmake could place the git-revision-version in a macro in a generated header or something like that, so that code could then either display it, or a calling-code could display it, etc.
Also, would it be possible for it to be displayed by cmake when we build openturbine?

rafmudaf replied:
@alanw0 Good point, I've created a pull request with this change in #11. As for the generated header with the git version, OpenTurbine is currently set up in this way following AMR Wind. The main CMakeList calls the cmake macro `generate_git_info` in [openturbine-utils.cmake](https://github.com/Exawind/openturbine/blob/main/cmake/openturbine-utils.cmake). That macro uses a few routines in [GetGitRevisionDescription.cmake](https://github.com/Exawind/openturbine/blob/main/cmake/GetGitRevisionDescription.cmake) to parse the git-status. Ultimately, OpenTurbineVersion.H is generated based on the results from [OpenTurbineVersion.H.in](https://github.com/Exawind/openturbine/blob/main/cmake/OpenTurbineVersion.H.in) in the [src/CMakeLists.txt](https://github.com/Exawind/openturbine/blob/main/src/CMakeLists.txt#L31). This header is currently only used to display the version info in the CLI and unit test driver:

```
$ ./openturbine -h
==============================================================================
                OpenTurbine (https://github.com/exawind/openturbine)

  OpenTurbine version :: v0.0.0-37-gcef9bff-DIRTY
  OpenTurbine Git SHA :: cef9bffc11001faba73520580ba129bc5d3a8ef7-DIRTY
  Exec. time       :: Thu Feb  9 13:56:51 2023

  No additional third-party libraries enabled

           This software is released under the MIT License.                
 See https://github.com/Exawind/openturbine/blob/main/LICENSE for details. 
------------------------------------------------------------------------------
```

alanw0 replied:
ok, great, sounds good.
I'm currently working on building openturbine myself, had to take a detour to clone/build/install kokkos...
Then I'll be much more familiar with all this stuff.

faisal-bhuiyan commented:
Updates to the PR:

> Create API documentation via Doxygen - Use of doxygen specific annotation embedded within C++ source code is pretty common to generate API documentation with minimal effort, e.g. in [AMR wind](https://exawind.github.io/amr-wind/developer/documentation.html#documenting-source-code).

- Doxygen setup files were for the project as part of this PR. The bulk of the changes made in this PR originate from this.

> Adopt a version number for project in CMakeLists.txt - we might consider adopting a [version number](https://cmake.org/cmake/help/latest/variable/CMAKE_PROJECT_VERSION.html) for OpenTurbine using CMake. We could use the version number along with a log file (discussed below) to quickly check the build version for development as well as for releases.

- Based on our discussion and subsequent changes implemented in PR#11, this work became redundant. 

> Create a logging utility - a logging utility for recording log messages to files (alongside the console) should be a very useful development tool and the first point of debugging whenever necessary. Various open-source libraries are available for this such as [spdlog](https://github.com/gabime/spdlog), however, we could just implement our own basic logging utility with minimal effort.

- Implemented a new feature for writing log messages using a Singleton design pattern. Replaced the use of `std::cout` with the log utility to print log messages to console and a text file. Unit tests were added as well.

> Adopt a C++ Style guide - We are currently using the C++ style guide in the form of a `clang-format` as recommended in [amr-wind](https://github.com/Exawind/amr-wind/blob/main/.clang-format). However, having a descriptive style guide helps to know about the thought process behind the choices made. The [Google C++ style guide](https://google.github.io/styleguide/cppguide.html) might be a good candidate to this end, of course with a few modifications of our own, e.g., increasing the [line length](https://google.github.io/styleguide/cppguide.html#Line_Length) from 80 characters to a more reasonable value such as 120.

- `clang-format` file was updated accordingly, along with comments to describe the why for each choice made.

In addition to the above, a slight adjustment was made to the folder structure within `src` which imo, adds more clarity to the project.

rafmudaf commented on src/main.cpp:
```diff
@@ -24,28 +27,45 @@ int main(int argc, char* argv[])
+    }
+    std::ofstream{log_file};
+
+#ifdef DEBUG
```
I wonder if this wouldn't be better to set as an input parameter. Whats the benefit to setting the severity level as a compiler flag? If it's an input parameter, you can enable/disable output for a given executable without having to go back and recompile.

faisal-bhuiyan replied:
My opinion regarding the above is we are taking care of setting the log severity level based on the build type provided to `CMake`. If someone's using the `Release` version, they probably don't care about the `Debug` messages being printed (since there should be far more `debug` messages than `info` messages which are valuable to only devs). Using the debug mode as input to the logging utility unintentionally could come with a big performance hit. Adding the compiler flags based on the build type takes care of this issue rather than adding one more (redundant imo) input to the executable and opening up avenues for error. Feel free to tell me if it doesn't make sense.

faisal-bhuiyan replied:
@rafmudaf I gave this a little more thought - I am open to the idea of setting the log level based on a parameter in the input file so it can be customized by the user. However, I think it will be useful mostly for `openturbine` developers and there exists such a switch for it in `CMakeLists.txt`. My opinion is to keep the current design and change things when there is such a feature request from users in the future.

alanw0 commented on src/heat/heat_solve.cpp:
```diff
@@ -73,20 +67,18 @@ double* heat_conduction_solver(
-    if (num < 0)
-        throw std::invalid_argument(
-            "linspace: num_in must be a positive integer");
+    auto start = static_cast<double>(start_in);
```
In general I think 'auto' makes it easier to write code, but the code is then harder to read. I think it's more important to make code easy to read. There are some good uses of auto, such as hiding ugly templating syntax which is basically language noise, such as std-library templated iterators etc. But usually I like to see the actual type rather than auto.
I guess in later c++ standards you can also use auto for function arguments, as kind of an alternative to templating. That's a separate issue.
In the code here, I think it would be more clear to simply write `double start = start_in;`. If the static_cast is safe, the assignment-conversion should be just as safe. (Unless it is doing a narrowing or sign-change that you want to explicitly do...)

faisal-bhuiyan replied:
@rafmudaf First of all, I apologize for not explaining my reasoning for refactoring the code. Please find it as follows, which should answer your question as well.

[It is considered good practice](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es11-use-auto-to-avoid-redundant-repetition-of-type-names) to use `auto` to avoid _redundant repetition_ of type names. For example

`double start = static_cast<double>(start_in);`

here the explicit type information on the left hand side is redundant (since it is clear we are casting it as a `double` from the rvalue), and we should prefer to use `auto` instead.

As a relevant aside, it's considered to be best practice to use [{}-initializer syntax for declaration with a named type](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es23-prefer-the--initializer-syntax) to avoid unwanted (and implicit) narrowing conversions. However, with auto, we tend to use `=` because there no potentially dangerous type conversion involved.

@alanw0  Based on the discussion [here](https://stackoverflow.com/questions/1609163/what-is-the-difference-between-static-cast-and-c-style-casting), looks like the C++ style casts are better than C-style assignment conversion since the former is checked by the compiler and expresses the intent of the programmer. I am open to your suggestion on this.

alanw0 replied:
I agree that auto is used to avoid redundant repetition of types. But note in the good practices document you linked, it says 'long hard-to-remember types', and also specifies an exception to avoid using auto for '... cases where you know exactly which type you want' and of course in cases where an unintended initializer-list might result. In the end I guess it's a judgment call. I definitely think the motivating priority is readability rather than writability. Many times the use of auto means that someone else reading the code will have to 'dive' or look somewhere else to find out what the type actually is.

Regarding the conversion, I definitely agree that C++ casts are better than C-style casts. I'm not as sure about assignment conversions. I think that if the conversion will result in narrowing or other things, the compiler will indeed tell you. (At least it will if you have warnings on.) If you want to be as explicit as possible, then I agree that static_cast provides that. In this case since `start_in` is a template parameter, and thus is not assured of being a convertible type, then the static_cast is probably a good idea.
A common example of static_cast is to do something like `unsigned i = static_cast<unsigned>(-1);` which is legal and correct, but I think it's much more clear to do this: `unsigned i = std::numeric_limits<unsigned>::max();`. Of course in both cases you could use auto at the beginning, but I think the repetition is less of an issue for primitive types.

faisal-bhuiyan replied:
> I definitely think the motivating priority is readability rather than writability.

Thanks for your comment, @alanw0 - I can definitely get behind your take above. Let's consider this comment resolved.

alanw0 replied:
Yes this is all good. Sorry for the delay in getting back to you. We are in good agreement. In practice if you use auto more than I would, I wouldn't hold up a pull-request over that.

faisal-bhuiyan replied:
Thanks for the review, @alanw0!

faisal-bhuiyan commented on CMakeLists.txt:
```diff
@@ -18,6 +18,39 @@ option(OTURB_ENABLE_ALL_WARNINGS "Show most warnings for most compilers" ON)
 # option(OTURB_ENABLE_FCOMPARE "Enable building fcompare when not testing" OFF)
 
+# Set RELEASE as default build type if none provided
+set(SUPPORTED_BUILD_TYPES None Debug Release RelWithDebInfo MinSizeRel)
```
The intent here is to check the value provided by the user and make sure it's within the supported values, as checked [here](https://github.com/Exawind/openturbine/pull/9/files/df411348c1cdef5d329d6cae7f46c496b1fa3de8#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR38). We are checking if the provided value is supported, and if not, the build type is set to default (accompanied with a warning message).

What's done in OpenFAST is similar, however, it's only checked if there's a value provided for the build type, not if the provided value makes sense.

faisal-bhuiyan commented on CMakeLists.txt:
```diff
@@ -18,6 +18,39 @@ option(OTURB_ENABLE_ALL_WARNINGS "Show most warnings for most compilers" ON)
 
+# Set RELEASE as default build type if none provided
+set(SUPPORTED_BUILD_TYPES None Debug Release RelWithDebInfo MinSizeRel)
+string(REPLACE ";" " " SUPPORTED_BUILD_TYPES_STRING "${SUPPORTED_BUILD_TYPES}")
```
> What would the "None" build type be?

Thanks for catching this, it remained from a previous design I was considering for this.

> Also, just heads up that the options provided are the defaults from CMake so this may be redundant.

Based on the reply to your previous comment on line 22, does the list make sense? I might be missing something here.
